File: JavascriptCommon/TextGenLib/TextGenLib.js
// Copyright(c) Lester J. Clark and Contributors.
// Licensed under the MIT License.
// TextGenLib.js
// <script src="../LJCCommon.js"></script>
// <script src="Script/RepeatItems.js"></script>
// <script src="Script/Replacements.js"></script>
// <script src="Script/Sections.js"></script>

// Generate output text from a template and data.
class TextGenLib
{
  // Checks if there are items.
  static HasItems(items)
  {
    let retValue = false;

    if (items != null
      && items.Count() > 0)
    {
      retValue = true;
    }
    return retValue;
  }

  // Private Properties
  #ActiveReplacements = [];
  #Lines = [];
  #Output = "";
  #Sections = [];

  // The Constructor function.
  constructor()
  {
    this.CommentChars = "//";
    this.PlaceholderBegin = "_";
    this.PlaceholderEnd = "_";
  }

  // Main Processing Methods

  // Generate the #Output text.
  TextGen(sections, templateLines)
  {
    this.#Sections = sections;
    this.#Lines = templateLines;

    let lineIndex = { Value: 0 };
    for (lineIndex.Value = 0; lineIndex.Value < templateLines.length
      ; lineIndex.Value++)
    {
      let line = this.#Lines[lineIndex.Value];
      if (this.#IsConfig(line))
      {
        continue;
      }

      let directive = Directive.GetDirective(line, this.CommentChars);
      if (directive != null
        && directive.IsSectionBegin())
      {
        let section = this.#Sections.Retrieve(directive.Name);
        if (null == section)
        {
          // No Section data.
          this.#SkipSection(lineIndex.Value);
        }
        else
        {
          lineIndex.Value++;
          section.BeginLineIndex = lineIndex.Value;
          this.#DoItems(section, lineIndex);
        }
        continue;
      }
      this.#AddOutput(line);
    }
    return this.#Output;
  }

  // Process the #IfBegin directive.
  #DoIf(directive, replacements, lineIndex)
  {
    // Check for #IfBegin.
    let success = true;
    let isIf = false;
    let isMatch = false;
    if (!LJC.HasText(directive.Value))
    {
      success = false;
    }

    // Check replacement value against directive value.
    if (success)
    {
      let replacement = replacements.Retrieve(directive.Name);
      if ("hasvalue" == directive.Value.toLowerCase())
      {
        isIf = true;
        if (replacement != null
          && LJC.HasText(replacement.Value))
        {
          isMatch = true;
        }
      }
      else
      {
        isIf = true;
        if (replacement != null
          && replacement.Value == directive.Value)
        {
          isMatch = true;
        }
      }
    }

    if (success)
    {
      // Process to IfEnd.
      lineIndex.Value++;
      for (let index = lineIndex.Value; index < this.#Lines.length; index++)
      {
        let line = this.#Lines[index];
        if (isMatch
          && !Directive.IsIfElse(line, this.CommentChars))
        {
          this.#DoOutput(replacements, line);
        }
        if (isIf
          && Directive.IsIfElse(line, this.CommentChars))
        {
          isMatch = !isMatch;
        }
        if (Directive.IsIfEnd(line, this.CommentChars))
        {
          lineIndex.Value = index;
          break;
        }
      }
    }
  }

  // Process the RepeatItems.
  #DoItems(section, lineIndex)
  {
    let success = true;
    let items = section.RepeatItems;

    // No Section data.
    if (!TextGenLib.HasItems(items))
    {
      success = false;
      lineIndex.Value = this.#SkipSection(lineIndex.Value);
      lineIndex++;
    }

    if (success)
    {
      for (let itemIndex = 0; itemIndex < items.Count(); itemIndex++)
      {
        let item = items.Items(itemIndex);

        // No Replacement data.
        if (!TextGenLib.HasItems(item.Replacements))
        {
          lineIndex.Value = this.#SkipSection(lineIndex.Value);

          // If not last item.
          if (itemIndex < items.Count() - 1)
          {
            // Do section again for following items.
            lineIndex.Value = section.BeginLineIndex;
          }
          continue;
        }

        for (let index = lineIndex.Value; index < this.#Lines.length; index++)
        {
          let line = this.#Lines[index];
          lineIndex.Value = index;

          let process = true;
          let directive = Directive.GetDirective(line, this.CommentChars);
          if (directive != null)
          {
            if (directive.IsSectionBegin())
            {
              let nextSection = this.#Sections.Retrieve(directive.Name);
              if (null == nextSection)
              {
                // No Section data.
                index = this.#SkipSection(lineIndex.Value);
                continue;
              }

              // RepeatItem processing starts with first line.
              lineIndex.Value++;
              nextSection.BeginLineIndex = lineIndex.Value;

              this.#AddActive(item);
              this.#DoItems(nextSection, lineIndex);
              this.#RemoveActive();

              // Continue with returned line index after the processed section.
              index = lineIndex.Value;
            }

            if (directive.IsSectionEnd())
            {
              // If not last item.
              if (itemIndex < items.Count() - 1)
              {
                // Do section again for following items.
                lineIndex.Value = section.BeginLineIndex;
                break;
              }
            }

            if (directive.IsIfBegin())
            {
              this.#DoIf(directive, item.Replacements, lineIndex);
            }
            index = lineIndex.Value;
          }

          if (process)
          {
            // Does not output directives.
            this.#DoOutput(item.Replacements, line);
          }
        }
      }
    }
  }

  // If not directive, process replacements and add to output.
  #DoOutput(replacements, line)
  {
    if (!Directive.IsDirective(line, this.CommentChars))
    {
      let lineItem = { Value: line };
      if (line != null
        && line.trim().length > 0)
      {
        this.#DoReplacements(replacements, lineItem);
      }
      this.#AddOutput(lineItem.Value);
    }
  }

  // Perform the line replacements.
  #DoReplacements(replacements, lineItem)
  {
    if (lineItem.Value.includes(this.PlaceholderBegin))
    {
      replacements.Sort();
      let line = lineItem.Value;

      //const matches = line.match(/(?<=_).+?(?=_)/g);
      let pattern = `${this.PlaceholderBegin}.+?${this.PlaceholderEnd}`;
      let matchRegex = new RegExp(pattern, "g");
      const matches = line.match(matchRegex);
      for (let index = 0; index < matches.length; index++)
      {
        let match = matches[index];
        let replacement = replacements.Retrieve(match);
        if (replacement != null)
        {
          lineItem.Value = lineItem.Value.replaceAll(match, replacement.Value);
        }
        else
        {
          // Replacement not found in current collection.
          // Search active replacements.
          let active = this.#ActiveReplacements;
          for (let activeIndex = active.length - 1; activeIndex >= 0; activeIndex--)
          {
            replacement = active[activeIndex].Retrieve(match);
            if (replacement != null)
            {
              lineItem.Value = lineItem.Value.replaceAll(match
                , replacement.Value);
              break;
            }
          }
        }
      }
    }
  }

  // Skips to the end of the current section.
  #SkipSection(lineIndexValue)
  {
    let retValue = lineIndexValue;

    // Skip to end of section.
    for (let index = lineIndexValue; index < this.#Lines.length; index++)
    {
      let line = this.#Lines[index];
      if (Directive.IsSectionEnd(line, this.CommentChars))
      {
        retValue = index++;
        break;
      }
    }
    return retValue;
  }

  // Other Methods

  // Add current replacements to Active array.
  #AddActive(item)
  {
    if (TextGenLib.HasItems(item.Replacements))
    {
      this.#ActiveReplacements.push(item.Replacements);
    }
  }

  // Add the line to the output.
  #AddOutput(line)
  {
    if (this.#Output.length > 0)
    {
      this.#Output += "\r\n";
    }
    this.#Output += line;
  }

  // Sets the configuration values.
  #IsConfig(line)
  {
    let retValue = false;

    // Sets the configuration.
    if (Directive.IsDirective(line, this.CommentChars))
    {
      var directive = Directive.GetDirective(line, this.CommentChars);
      switch (directive.ID.toLowerCase())
      {
        case "#commentchars":
          retValue = true;
          this.CommentChars = directive.Name;
          break;

        case "#placeholderbegin":
          retValue = true;
          this.PlaceholderBegin = directive.Name;
          break;

        case "#placeholderend":
          retValue = true;
          this.PlaceholderEnd = directive.Name;
          break;
      }
    }
    return retValue;
  }

  // Remove Replacements that are no longer active.
  #RemoveActive()
  {
    if (this.#ActiveReplacements.length > 0)
    {
      this.#ActiveReplacements.pop();
    }
  }
}

// Represents a Directive object.
class Directive
{
  // static check Line Functions

  // Checks line for directive and returns the directive object.
  // Returns directive if line is a directive.
  static GetDirective(line, commentChars)
  {
    let retValue = null;

    // Directive Layout = "// #Directive Name [DataType]"
    // Checks for #CommentChars as property may not be set.
    if (line != null
      && (line.trim().startsWith(commentChars)
        || line.toLowerCase().includes("#commentchars")))
    {
      let tokens = line.trim().split(/\s+/g);
      if (tokens.length > 1
        && tokens[1].startsWith("#"))
      {
        let id = tokens[1];
        let name = null;
        if (tokens.length > 2)
        {
          name = tokens[2];
        }
        let value = null;
        if (tokens.length > 3)
        {
          value = tokens[3];
        }
        retValue = new Directive(id, name, value);
      }
    }
    return retValue;
  }

  // Checks if the line has a directive.
  static IsDirective(line, commentChars)
  {
    let retValue = false;

    let directive = Directive.GetDirective(line, commentChars);
    if (directive != null)
    {
      let lowerName = directive.ID.toLowerCase();
      if (lowerName == "#commentchars"
        || lowerName == "#placeholderbegin"
        || lowerName == "#placeholderend"
        || lowerName == "#sectionbegin"
        || lowerName == "#sectionend"
        || lowerName == "#value"
        || lowerName == "#ifbegin"
        || lowerName == "#ifend")
      {
        retValue = true;
      }
    }
    return retValue;
  }

  // Checks if the line is a SectionEnd.  // 
  static IsIfBegin(line, commentChars)
  {
    let retValue = false;

    // Directive Layout = "// #SectionEnd Name"
    let directive = Directive.GetDirective(line, commentChars);
    if (directive != null
      && "#ifbegin" == directive.ID.toLowerCase())
    {
      retValue = true;
    }
    return retValue;
  }

  // Checks if the line is a SectionEnd.  // 
  static IsIfElse(line, commentChars)
  {
    let retValue = false;

    // Directive Layout = "// #SectionEnd Name"
    let directive = Directive.GetDirective(line, commentChars);
    if (directive != null
      && "#ifelse" == directive.ID.toLowerCase())
    {
      retValue = true;
    }
    return retValue;
  }

  // Checks if the line is a SectionEnd.  // 
  static IsIfEnd(line, commentChars)
  {
    let retValue = false;

    // Directive Layout = "// #SectionEnd Name"
    let directive = Directive.GetDirective(line, commentChars);
    if (directive != null
      && "#ifend" == directive.ID.toLowerCase())
    {
      retValue = true;
    }
    return retValue;
  }

  // Checks if the line is a SectionBegin.
  static IsSectionBegin(line, commentChars)
  {
    let retValue = false;

    // Directive Layout = "// #SectionBegin Name"
    let directive = Directive.GetDirective(line, commentChars);
    if (directive != null
      && "#sectionbegin" == directive.ID.toLowerCase())
    {
      retValue = true;
    }
    return retValue;
  }

  // Checks if the line is a SectionEnd.  // 
  static IsSectionEnd(line, commentChars)
  {
    let retValue = false;

    // Directive Layout = "// #SectionEnd Name"
    let directive = Directive.GetDirective(line, commentChars);
    if (directive != null
      && "#sectionend" == directive.ID.toLowerCase())
    {
      retValue = true;
    }
    return retValue;
  }

  // The Constructor method.
  constructor(id, name, value = null)
  {
    this.ID = id;
    this.Name = name;
    this.Value = value;
  }

  // Public Methods

  // Checks if directive ID = IfBegin.
  IsIfBegin()
  {
    let retValue = false;

    if ("#ifbegin" == this.ID.toLowerCase())
    {
      retValue = true;
    }
    return retValue;
  }

  // Checks if directive ID = IfElse.
  IsIfElse()
  {
    let retValue = false;

    if ("#ifelse" == this.ID.toLowerCase())
    {
      retValue = true;
    }
    return retValue;
  }

  // Checks if directive ID = IfEnd.
  IsIfEnd()
  {
    let retValue = false;

    if ("#ifend" == this.ID.toLowerCase())
    {
      retValue = true;
    }
    return retValue;
  }

  // Checks if directive ID = SectionBegin.
  IsSectionBegin()
  {
    let retValue = false;

    if ("#sectionbegin" == this.ID.toLowerCase())
    {
      retValue = true;
    }
    return retValue;
  }

  // Checks if directive ID = SectionEnd.
  IsSectionEnd()
  {
    let retValue = false;

    if ("#sectionend" == this.ID.toLowerCase())
    {
      retValue = true;
    }
    return retValue;
  }
}

File: CoreUtilities/LJCGenText/LJCGenTextLib/TextGenLib.cs
// Copyright(c) Lester J. Clark and Contributors.
// Licensed under the MIT License.
// TextGenLib.cs
using LJCNetCommon;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace LJCGenTextLib
{
  /// <summary>Generate output text from a template and data.</summary>
  public class TextGenLib
  {
    /// <summary>Initializes an object instance.</summary>
    public TextGenLib()
    {
      CommentChars = "//";
      PlaceholderBegin = "_";
      PlaceholderEnd = "_";

      // Private Properties
      ActiveReplacements = new List<Replacements>();
      Output = "";
    }

    #region Main Processing Methods

    // Generate the Output text.
    /// <include path='items/TextGen/*' file='Doc/TextGenLib.xml'/>
    public string TextGen(Sections sections, string[] templateLines)
    {
      Sections = sections;
      Lines = templateLines;

      for (var lineIndex = 0; lineIndex < templateLines.Length; lineIndex++)
      {
        var line = Lines[lineIndex];
        if (IsConfig(line, out Directive directive))
        {
          continue;
        }

        // Only SectionBegin is valid. Other directives should not appear.
        if (directive != null)
        {
          var name = directive.ID;
          if (name.ToLower() != "#sectionbegin")
          {
            continue;
          }
        }

        if (directive != null
          && directive.IsSectionBegin())
        {
          var section = Sections.Retrieve(directive.Name);
          if (null == section)
          {
            // No Section data.
            lineIndex = SkipSection(lineIndex, directive.Name);
          }
          else
          {
            lineIndex++;
            section.BeginLineIndex = lineIndex;
            DoItems(section, ref lineIndex);
            continue;
          }
        }

        // Add text before first section and after last section.
        AddOutput(line);
      }
      return Output;
    }

    // Performs replacement from active replacement value.
    private string ActiveReplacement(string line, string replacementName)
    {
      string retValue = line;

      var value = ActiveValue(replacementName);
      if (NetString.HasValue(value))
      {
        retValue = retValue.Replace(replacementName, value);
      }
      return retValue;
    }

    // Gets the active replacement value.
    private string ActiveValue(string replacementName)
    {
      string retValue = null;

      var active = ActiveReplacements;
      for (var index = active.Count - 1; index >= 0; index--)
      {
        var replacement = active[index].Retrieve(replacementName);
        if (replacement != null)
        {
          retValue = replacement.Value;
          break;
        }
      }
      return retValue;
    }

    // Process the #IfBegin directive.
    private void DoIf(Directive directive, Replacements replacements
      , ref int lineIndex)
    {
      // Check replacement value against directive value.
      var success = true;
      var isIf = false;
      var isMatch = false;
      if (!NetString.HasValue(directive.Value))
      {
        success = false;
      }

      // Check replacement value against directive value.
      if (success)
      {
        isIf = true;

        string value;
        var replacement = replacements.Retrieve(directive.Name);
        if (replacement != null)
        {
          value = replacement.Value;
        }
        else
        {
          value = ActiveValue(directive.Name);
        }

        if (NetString.HasValue(value))
        {
          if ("hasvalue" == directive.Value.ToLower()
            || value == directive.Value)
          {
            isMatch = true;
          }
        }
      }

      if (success)
      {
        // Process to IfEnd.
        lineIndex++;
        for (var index = lineIndex; index < Lines.Length; index++)
        {
          var line = Lines[index];
          if (isMatch
            && !Directive.IsIfElse(line, CommentChars))
          {
            DoOutput(replacements, line);
          }
          if (isIf
            && Directive.IsIfElse(line, CommentChars))
          {
            isMatch = !isMatch;
          }
          if (Directive.IsIfEnd(line, CommentChars))
          {
            lineIndex = index;
            break;
          }
        }
      }
    }

    // Process the RepeatItems.
    private void DoItems(Section section, ref int nextLineIndex)
    {
      var success = true;
      var repeatItems = section.RepeatItems;

      // No Section data.
      if (!NetCommon.HasItems(repeatItems))
      {
        success = false;
        nextLineIndex = SkipSection(nextLineIndex, section.Name);
        nextLineIndex++;
      }

      if (success)
      {
        // Process each RepeatItem.
        for (var itemIndex = 0; itemIndex < repeatItems.Count; itemIndex++)
        {
          var repeatItem = repeatItems[itemIndex];

          // No Replacement data.
          if (!NetCommon.HasItems(repeatItem.Replacements))
          {
            nextLineIndex = SkipSection(nextLineIndex, section.Name);

            // If not last repeat item.
            if (itemIndex < repeatItems.Count - 1)
            {
              // Do section again for following items.
              nextLineIndex = section.BeginLineIndex;
            }
            continue;
          }

          // Process template lines.
          for (var lineIndex = nextLineIndex; lineIndex < Lines.Length; lineIndex++)
          {
            var line = Lines[lineIndex];
            nextLineIndex = lineIndex;

            var directive = Directive.GetDirective(line, CommentChars);

            // Skip unprocessed directives.
            if (directive != null
              && IsNotProcessValues(directive.ID))
            {
              continue;
            }

            // Do output if directive is null or is not a processed directive.
            if (directive == null
              || (directive != null
              && !IsProcessValues(directive.ID)))
            {
              DoOutput(repeatItem.Replacements, line);
              continue;
            }

            if (directive.IsSectionBegin())
            {
              if (DoSubsection(directive, repeatItem, ref nextLineIndex))
              {
                lineIndex = nextLineIndex;
                continue;
              }

              var nextSection = GetBeginSection(line);
              if (null == nextSection)
              {
                // No Section data.
                lineIndex = SkipSection(nextLineIndex, directive.Name);
                continue;
              }

              ProcessItems(nextSection, repeatItem, ref nextLineIndex);
              lineIndex = nextLineIndex;
              continue;
            }

            if (directive.IsSectionEnd())
            {
              // If not last item.
              if (itemIndex < repeatItems.Count - 1)
              {
                // Do section again for following items.
                nextLineIndex = section.BeginLineIndex;
              }
              break;
            }

            if (directive.IsIfBegin())
            {
              DoIf(directive, repeatItem.Replacements, ref nextLineIndex);
              lineIndex = nextLineIndex;
            }
          }
        }
      }
    }

    // If not directive, process replacements and add to output.
    private string DoOutput(Replacements replacements, string line)
    {
      string retValue = null;

      if (!Directive.IsDirective(line, CommentChars))
      {
        if (line != null
          && line.Trim().Length > 0)
        {
          DoReplacements(replacements, ref line);
        }
        retValue = AddOutput(line);
      }
      return retValue;
    }

    // Perform the line replacements.
    private void DoReplacements(Replacements replacements
      , ref string lineItem)
    {
      if (lineItem.Contains(PlaceholderBegin))
      {
        replacements.Sort();
        var line = lineItem;

        var pattern = $"{PlaceholderBegin}.+?{PlaceholderEnd}";
        var matches = Regex.Matches(line, pattern);
        for (var index = 0; index < matches.Count; index++)
        {
          var match = matches[index];
          var replacement = replacements.Retrieve(match.Value);
          if (replacement != null)
          {
            lineItem = lineItem.Replace(match.Value, replacement.Value);
          }
          else
          {
            lineItem = ActiveReplacement(lineItem, match.Value);
          }
        }
      }
    }

    private bool DoSubsection(Directive directive, RepeatItem repeatItem
      , ref int nextLineIndex)
    {
      bool retValue = false;

      if (directive.Name.ToLower() == "subsection"
        && repeatItem.Subsection != null)
      {
        var subSection = repeatItem.Subsection;
        if (subSection != null)
        {
          ProcessItems(subSection, repeatItem, ref nextLineIndex);
          retValue = true;
        }
      }
      return retValue;
    }

    // Process the repeat items.
    private void ProcessItems(Section section, RepeatItem repeatItem
      , ref int nextLineIndex)
    {
      // RepeatItem processing starts with first line.
      nextLineIndex++;
      section.BeginLineIndex = nextLineIndex;

      AddActive(repeatItem);
      DoItems(section, ref nextLineIndex);
      RemoveActive();
    }

    // Skips to the end of the current section.
    private int SkipSection(int lineIndex, string name)
    {
      var retValue = lineIndex;

      // Skip to end of section.
      for (var index = lineIndex; index < Lines.Length; index++)
      {
        var line = Lines[index];

        var directive = Directive.GetDirective(line, CommentChars);
        if (Directive.IsSectionEnd(line, CommentChars)
          && directive.Name == name)
        {
          retValue = index++;
          break;
        }
      }
      return retValue;
    }
    #endregion

    #region Other Methods

    // Add current replacements to Active array.
    private void AddActive(RepeatItem item)
    {
      if (NetCommon.HasItems(item.Replacements))
      {
        ActiveReplacements.Add(item.Replacements);
      }
    }

    // Add the line to the output.
    private string AddOutput(string line)
    {
      string retValue = null;

      if (!line.Trim().StartsWith("<!--X"))
      {
        if (Output.Length > 0)
        {
          Output += "\r\n";
        }
        Output += line;
        retValue = line;
      }
      return retValue;
    }

    // Gets the begin section.
    private Section GetBeginSection(string line)
    {
      var directive = Directive.GetDirective(line, CommentChars);
      var retValue = Sections.Retrieve(directive.Name);
      return retValue;
    }

    // Sets the configuration values.
    private bool IsConfig(string line, out Directive directive)
    {
      var retValue = false;

      // Sets the configuration.
      if (line.ToLower().Contains("#commentchars"))
      {
        string[] values = NetString.Split(line, " ");
        if (values.Length >= 2)
        {
          CommentChars = values[2];
        }
      }

      directive = Directive.GetDirective(line, CommentChars);
      if (directive != null)
      {
        switch (directive.ID.ToLower())
        {
          case "#commentchars":
            retValue = true;
            CommentChars = directive.Name;
            break;

          case "#placeholderbegin":
            retValue = true;
            PlaceholderBegin = directive.Name;
            break;

          case "#placeholderend":
            retValue = true;
            PlaceholderEnd = directive.Name;
            break;
        }
      }
      return retValue;
    }

    // Checks if a value is in a list of names.
    private bool IsInValues(string text, params string[] values)
    {
      bool retValue = false;

      foreach (var value in values)
      {
        if (value.ToLower() == text.ToLower())
        {
          retValue = true;
          break;
        }
      }
      return retValue;
    }

    // true if not DoItems process values; otherwise false.
    private bool IsNotProcessValues(string name)
    {
      var retValue = IsInValues(name, "#Value", "#IfElse", "#IfEnd");
      return retValue;
    }

    // true if DoItems process values; otherwise false.
    private bool IsProcessValues(string name)
    {
      var retValue = IsInValues(name, "#SectionBegin", "#SectionEnd"
        , "#IfBegin");
      return retValue;
    }

    // Remove Replacements that are no longer active.
    private void RemoveActive()
    {
      if (NetCommon.HasItems(ActiveReplacements))
      {
        ActiveReplacements.RemoveAt(ActiveReplacements.Count - 1);
      }
    }
    #endregion

    #region Properties

    /// <summary></summary>
    public string CommentChars { get; set; }

    /// <summary></summary>
    public string PlaceholderBegin { get; set; }

    /// <summary></summary>
    public string PlaceholderEnd { get; set; }

    private List<Replacements> ActiveReplacements { get; set; }

    private string[] Lines { get; set; }

    private string Output { get; set; }

    private Sections Sections { get; set; }
    #endregion
  }
}